pwndbg/heap_tracking.py

```diff
@@ -0,0 +1,357 @@
+            for i in range(lo_i, hi_i):
+                addr, ch = self.free_chunks.popitem(index=i)
+
+                self.free_wps[addr].delete()
+                del self.free_wps[addr]
+
+            # Add new handlers in their place. We scan over all of the chunks in
+            # the heap in the range of affected chunks, and add the ones that
+            # are free.
+            allocator = pwndbg.heap.current
+            bins_list = [
+                allocator.fastbins(lo_heap.arena.address),
+                allocator.smallbins(lo_heap.arena.address),
+                allocator.largebins(lo_heap.arena.address),
+                allocator.unsortedbin(lo_heap.arena.address),
+            ]
+            if allocator.has_tcache():
+                bins_list.append(allocator.tcachebins(None))
+            bins_list = [x for x in bins_list if x is not None]
+
+            print(f"Chunks in heap {lo_heap.start:#x}-{lo_heap.end:#x}:")
+            for ch in lo_heap:
+                # Check for range overlap.
+                ch_lo_addr = ch.address
+                ch_hi_addr = ch.address + ch.size
+
+                print(f"        {ch.address:#x} ({ch.real_size} bytes)", end="")
+
+                lo_in_range = ch_lo_addr < hi_addr
+                hi_in_range = ch_hi_addr > lo_addr
+
+                if not lo_in_range or not hi_in_range:
+                    # No range overlap.
+                    print(" (out of range)")
+                    continue
+
+                # Check if the chunk is free.
+                found = False
+                for b in bins_list:
+                    if b.contains_chunk(ch.real_size, ch.address):
+                        # The chunk is free. Add it to the free list and install
+                        # a new watch point for it.
+                        nch = Chunk(ch.address, ch.size, ch.real_size, 0)
+                        wp = FreeChunkWatchpoint(nch, self)
+
+                        print(" (free)")
+
+                        self.free_chunks[ch.address] = nch
+                        self.free_wps[ch.address] = wp
+
+                        # Move on to the next chunk.
+                        found = True
+                        break
+                if not found:
+                    print(" (in use)")
+
+        self.alloc_chunks[chunk.address] = chunk
+
+    def free(self, address) -> bool:
+        if address not in self.alloc_chunks:
+            return False
+        chunk = self.alloc_chunks[address]
+        wp = FreeChunkWatchpoint(chunk, self)
+
+        self.free_chunks[chunk.address] = chunk
+        self.free_wps[chunk.address] = wp
+
+        return True
+
+
+class MallocEnterBreakpoint(gdb.Breakpoint):
+    def __init__(self, tracker):
+        super().__init__(MALLOC_NAME, internal=True)
+        self.tracker = tracker
+
+    def stop(self):
+        pwndbg.gdblib.regs.__getattr__.cache.clear()
+        requested_size = pwndbg.arguments.argument(0)
```

Comment:
on 32-bit the arg may be on stack and so clearing regs cache may probably not be enough if its cached? not sure if it is.

Reply:
It should be enough, but, as discussed above, all of the stop caches are cleared now.